File: Projetos/ACBrLib/Demos/C#/Mail/Imports/Dinamico/MT/ACBrMail.cs
using System;
using System.Text;
using ACBrLib.Core;
using ACBrLib.Core.Mail;

namespace ACBrLib.Mail
{
    public sealed partial class ACBrMail : ACBrLibHandle
    {
        #region Constructors

        public ACBrMail(string eArqConfig = "", string eChaveCrypt = "") : base(IsWindows ? "ACBrMail64.dll" : "libacbrmail64.so",
                                                                                IsWindows ? "ACBrMail32.dll" : "libacbrmail32.so")
        {
            var inicializar = GetMethod<MAIL_Inicializar>();
            var ret = ExecuteMethod(() => inicializar(ref libHandle, ToUTF8(eArqConfig), ToUTF8(eChaveCrypt)));

            CheckResult(ret);
            Config = new MailConfig(this);
        }

        #endregion Constructors

        #region Properties

        public string Nome
        {
            get
            {
                var bufferLen = BUFFER_LEN;
                var buffer = new StringBuilder(bufferLen);

                var method = GetMethod<MAIL_Nome>();
                var ret = ExecuteMethod(() => method(libHandle, buffer, ref bufferLen));

                CheckResult(ret);

                return ProcessResult(buffer, bufferLen);
            }
        }

        public string Versao
        {
            get
            {
                var bufferLen = BUFFER_LEN;
                var buffer = new StringBuilder(bufferLen);

                var method = GetMethod<MAIL_Versao>();
                var ret = ExecuteMethod(() => method(libHandle, buffer, ref bufferLen));

                CheckResult(ret);

                return ProcessResult(buffer, bufferLen);
            }
        }

        public MailConfig Config { get; }

        #endregion Properties

        #region Methods

        #region Ini

        public override void ConfigGravar(string eArqConfig = "")
        {
            var gravarIni = GetMethod<MAIL_ConfigGravar>();
            var ret = ExecuteMethod(() => gravarIni(libHandle, ToUTF8(eArqConfig)));

            CheckResult(ret);
        }

        public override void ConfigLer(string eArqConfig = "")
        {
            var lerIni = GetMethod<MAIL_ConfigLer>();
            var ret = ExecuteMethod(() => lerIni(libHandle, ToUTF8(eArqConfig)));

            CheckResult(ret);
        }

        public override T ConfigLerValor<T>(ACBrSessao eSessao, string eChave)
        {
            var method = GetMethod<MAIL_ConfigLerValor>();

            var bufferLen = BUFFER_LEN;
            var pValue = new StringBuilder(bufferLen);
            var ret = ExecuteMethod(() => method(libHandle, ToUTF8(eSessao.ToString()), ToUTF8(eChave), pValue, ref bufferLen));
            CheckResult(ret);

            var value = ProcessResult(pValue, bufferLen);
            return ConvertValue<T>(value);
        }

        public override void ConfigGravarValor(ACBrSessao eSessao, string eChave, object value)
        {
            if (value == null) return;

            var method = GetMethod<MAIL_ConfigGravarValor>();
            var propValue = ConvertValue(value);

            var ret = ExecuteMethod(() => method(libHandle, ToUTF8(eSessao.ToString()), ToUTF8(eChave), ToUTF8(propValue)));
            CheckResult(ret);
        }

        public override void ImportarConfig(string eArqConfig = "")
        {
            var importarConfig = GetMethod<MAIL_ConfigImportar>();
            var ret = ExecuteMethod(() => importarConfig(libHandle, ToUTF8(eArqConfig)));

            CheckResult(ret);
        }

        public override string ExportarConfig()
        {
            var bufferLen = BUFFER_LEN;
            var buffer = new StringBuilder(bufferLen);

            var method = GetMethod<MAIL_ConfigExportar>();
            var ret = ExecuteMethod(() => method(libHandle, buffer, ref bufferLen));

            CheckResult(ret);

            return ProcessResult(buffer, bufferLen);
        }

        #endregion Ini

        public void SetSubject(string subject)
        {
            var method = GetMethod<MAIL_SetSubject>();
            var ret = ExecuteMethod(() => method(libHandle, ToUTF8(subject)));

            CheckResult(ret);
        }

        public void AddAddress(string eEmail, string eName)
        {
            var method = GetMethod<MAIL_AddAddress>();
            var ret = ExecuteMethod(() => method(libHandle, ToUTF8(eEmail), ToUTF8(eName)));

            CheckResult(ret);
        }

        public void AddReplyTo(string eEmail, string eName)
        {
            var method = GetMethod<MAIL_AddReplyTo>();
            var ret = ExecuteMethod(() => method(libHandle, ToUTF8(eEmail), ToUTF8(eName)));

            CheckResult(ret);
        }

        public void AddCC(string eEmail, string eName)
        {
            var method = GetMethod<MAIL_AddCC>();
            var ret = ExecuteMethod(() => method(libHandle, ToUTF8(eEmail), ToUTF8(eName)));

            CheckResult(ret);
        }

        public void AddBCC(string eEmail, string eName)
        {
            var method = GetMethod<MAIL_AddBCC>();
            var ret = ExecuteMethod(() => method(libHandle, ToUTF8(eEmail)));

            CheckResult(ret);
        }

        public void AddAttachment(string eFileName, string eDescription, MailAttachmentDisposition aDisposition)
        {
            var method = GetMethod<MAIL_AddAttachment>();
            var ret = ExecuteMethod(() => method(libHandle, ToUTF8(eFileName), ToUTF8(eDescription), (int)aDisposition));

            CheckResult(ret);
        }

        public void ClearAttachment()
        {
            var method = GetMethod<MAIL_ClearAttachment>();
            var ret = ExecuteMethod(() => method(libHandle));

            CheckResult(ret);
        }

        public void AddBody(string eBody)
        {
            var method = GetMethod<MAIL_AddBody>();
            var ret = ExecuteMethod(() => method(libHandle, ToUTF8(eBody)));

            CheckResult(ret);
        }

        public void AddAltBody(string eAltBody)
        {
            var method = GetMethod<MAIL_AddAltBody>();
            var ret = ExecuteMethod(() => method(libHandle, ToUTF8(eAltBody)));

            CheckResult(ret);
        }

        public void SaveToFile(string eFileName)
        {
            var method = GetMethod<MAIL_SaveToFile>();
            var ret = ExecuteMethod(() => method(libHandle, ToUTF8(eFileName)));

            CheckResult(ret);
        }

        public void Clear()
        {
            var method = GetMethod<MAIL_Clear>();
            var ret = ExecuteMethod(() => method(libHandle));

            CheckResult(ret);
        }

        public void Send()
        {
            var method = GetMethod<MAIL_Send>();
            var ret = ExecuteMethod(() => method(libHandle));

            CheckResult(ret);
        }

        #region Private Methods

        protected override void FinalizeLib()
        {
            var finalizar = GetMethod<MAIL_Finalizar>();
            var codRet = ExecuteMethod(() => finalizar(libHandle));
            CheckResult(codRet);
        }

        protected override string GetUltimoRetorno(int iniBufferLen = 0)
        {
            var bufferLen = iniBufferLen < 1 ? BUFFER_LEN : iniBufferLen;
            var buffer = new StringBuilder(bufferLen);
            var ultimoRetorno = GetMethod<MAIL_UltimoRetorno>();

            if (iniBufferLen < 1)
            {
                ExecuteMethod(() => ultimoRetorno(libHandle, buffer, ref bufferLen));
                if (bufferLen <= BUFFER_LEN) return FromUTF8(buffer);

                buffer.Capacity = bufferLen;
            }

            ExecuteMethod(() => ultimoRetorno(libHandle, buffer, ref bufferLen));
            return FromUTF8(buffer);
        }

        #endregion Private Methods

        #endregion Methods
    }
}

File: Projetos/ACBrLib/Demos/Java/Mail/Imports/ACBrLibMailMT/src/com/acbr/mail/ACBrMail.java
package com.acbr.mail;

import com.acbr.ACBrLibBase;
import com.sun.jna.Library;
import com.sun.jna.Native;
import com.sun.jna.Platform;
import com.sun.jna.ptr.IntByReference;
import java.io.File;
import com.acbr.ACBrSessao;
import com.sun.jna.Pointer;
import com.sun.jna.ptr.PointerByReference;
import java.nio.ByteBuffer;
import java.nio.file.Paths;

public final class ACBrMail extends ACBrLibBase implements AutoCloseable  {    
       
    private interface ACBrMailLib extends Library {
        
        static String JNA_LIBRARY_NAME = LibraryLoader.getLibraryName();
        public final static ACBrMailLib INSTANCE = LibraryLoader.getInstance();
        
        class LibraryLoader {
            private static String library = "";
            private static ACBrMailLib instance = null;
            
            public static String getLibraryName() {
                if (library.isEmpty()) {
                    library = Platform.is64Bit() ? "ACBrMail64" : "ACBrMail32";
                }
                
                return library;
            }
            
            public static ACBrMailLib getInstance() {
                if (instance == null) {
                    instance = (ACBrMailLib) Native.synchronizedLibrary((Library) Native.loadLibrary(JNA_LIBRARY_NAME, ACBrMailLib.class));
                }
                
                return instance;
            }
        }
        
        int MAIL_Inicializar(PointerByReference libHandle, String eArqConfig, String eChaveCrypt);
        
        int MAIL_Finalizar(Pointer libHandler);
        
        int MAIL_Nome(Pointer libHandler, ByteBuffer buffer, IntByReference bufferSize);
        
        int MAIL_Versao(Pointer libHandler, ByteBuffer buffer, IntByReference bufferSize);
        
        int MAIL_UltimoRetorno(Pointer libHandler, ByteBuffer buffer, IntByReference bufferSize);
        
        int MAIL_ConfigImportar(Pointer libHandler, String eArqConfig);
        
	    int MAIL_ConfigExportar(Pointer libHandler, ByteBuffer buffer, IntByReference bufferSize);
        
        int MAIL_ConfigLer(Pointer libHandler, String eArqConfig);
        
        int MAIL_ConfigGravar(Pointer libHandler, String eArqConfig);
        
        int MAIL_ConfigLerValor(Pointer libHandler, String eSessao, String eChave, ByteBuffer buffer, IntByReference bufferSize);
        
        int MAIL_ConfigGravarValor(Pointer libHandler, String eSessao, String eChave, String valor);
          
        int MAIL_SetSubject(Pointer libHandler, String eSubject);  
        
        int MAIL_AddAddress(Pointer libHandler, String eEmail, String eName);
        
        int MAIL_AddReplyTo(Pointer libHandler, String eEmail, String eName);
        
        int MAIL_AddCC(Pointer libHandler, String eEmail, String eName);
        
        int MAIL_AddBCC(Pointer libHandler, String eEmail);
        
        int MAIL_ClearAttachment(Pointer libHandler);
        
        int MAIL_AddAttachment(Pointer libHandler, String eFileName, String eDescription, int aDisposition);
        
        int MAIL_AddBody(Pointer libHandler, String eBody);
        
        int MAIL_AddAltBody(Pointer libHandler, String eAltBody);
        
        int MAIL_SaveToFile(Pointer libHandler, String eFileName);
        
        int MAIL_Clear(Pointer libHandler);
        
        int MAIL_Send(Pointer libHandler);
    }
 
    public ACBrMail() throws Exception {
        File iniFile = Paths.get(System.getProperty("user.dir"), "ACBrLib.ini").toFile();
        if (!iniFile.exists()) {
            iniFile.createNewFile();
        }
        
        PointerByReference handle = new PointerByReference();
        int ret = ACBrMailLib.INSTANCE.MAIL_Inicializar(handle, toUTF8(iniFile.getAbsolutePath()), toUTF8(""));
        checkResult(ret);
    }

    public ACBrMail(String eArqConfig, String eChaveCrypt) throws Exception {
        PointerByReference handle = new PointerByReference();
        int ret = ACBrMailLib.INSTANCE.MAIL_Inicializar(handle, toUTF8(eArqConfig), toUTF8(eChaveCrypt));
        checkResult(ret);
    }
    
    @Override
    public void close() throws Exception {
        int ret = ACBrMailLib.INSTANCE.MAIL_Finalizar(getHandle());
        checkResult(ret);
    }
    
    @Override
    protected void finalize() throws Throwable {
        try {
            int ret = ACBrMailLib.INSTANCE.MAIL_Finalizar(getHandle());
            checkResult(ret);
        } finally {
            super.finalize();
        }
    }
    
    public String nome() throws Exception {
        ByteBuffer buffer = ByteBuffer.allocate(STR_BUFFER_LEN);
        IntByReference bufferLen = new IntByReference(STR_BUFFER_LEN);

        int ret = ACBrMailLib.INSTANCE.MAIL_Nome(getHandle(), buffer, bufferLen);
        checkResult(ret);

        return fromUTF8(buffer, bufferLen.getValue());
    }

    public String versao() throws Exception {
        ByteBuffer buffer = ByteBuffer.allocate(STR_BUFFER_LEN);
        IntByReference bufferLen = new IntByReference(STR_BUFFER_LEN);

        int ret = ACBrMailLib.INSTANCE.MAIL_Versao(getHandle(), buffer, bufferLen);
        checkResult(ret);

        return fromUTF8(buffer, bufferLen.getValue());
    }

    public void configLer() throws Exception {
        configLer("");
    }

    public void configLer(String eArqConfig) throws Exception {
        int ret = ACBrMailLib.INSTANCE.MAIL_ConfigLer(getHandle(), toUTF8(eArqConfig));
        checkResult(ret);
    }

    public void configGravar() throws Exception {
        configGravar("");
    }

    public void configGravar(String eArqConfig) throws Exception {
        int ret = ACBrMailLib.INSTANCE.MAIL_ConfigGravar(getHandle(), toUTF8(eArqConfig));
        checkResult(ret);
    }

    public String configLerValor(ACBrSessao eSessao, String eChave) throws Exception {
        ByteBuffer buffer = ByteBuffer.allocate(STR_BUFFER_LEN);
        IntByReference bufferLen = new IntByReference(STR_BUFFER_LEN);

        int ret = ACBrMailLib.INSTANCE.MAIL_ConfigLerValor(getHandle(), toUTF8(eSessao.name()), toUTF8(eChave), buffer, bufferLen);
        checkResult(ret);
        
        return processResult(buffer, bufferLen);
    }

    public void configGravarValor(ACBrSessao eSessao, String eChave, Object value) throws Exception {
        int ret = ACBrMailLib.INSTANCE.MAIL_ConfigGravarValor(getHandle(), toUTF8(eSessao.name()), toUTF8(eChave), toUTF8(value.toString()));
        checkResult(ret);
    }
    
    public void setSubject(String eMail) throws Exception {
        int ret = ACBrMailLib.INSTANCE.MAIL_SetSubject(getHandle(), toUTF8(eMail));
        checkResult(ret);
    }
    
    public void addAddress(String eMail, String eName) throws Exception {
        int ret = ACBrMailLib.INSTANCE.MAIL_AddAddress(getHandle(), toUTF8(eMail), toUTF8(eName));
        checkResult(ret);
    }
    
    public void addReplyTo(String eMail, String eName) throws Exception {
        int ret = ACBrMailLib.INSTANCE.MAIL_AddReplyTo(getHandle(), toUTF8(eMail), toUTF8(eName));
        checkResult(ret);
    }
    
    public void addCC(String eMail, String eName) throws Exception {
        int ret = ACBrMailLib.INSTANCE.MAIL_AddCC(getHandle(), toUTF8(eMail), toUTF8(eName));
        checkResult(ret);
    }
    
    public void addBCC(String eMail) throws Exception {
        int ret = ACBrMailLib.INSTANCE.MAIL_AddBCC(getHandle(), toUTF8(eMail));
        checkResult(ret);
    }
    
    public void clearAttachment() throws Exception {
        int ret = ACBrMailLib.INSTANCE.MAIL_ClearAttachment(getHandle());
        checkResult(ret);
    }
    
    public void addAttachment(String eFileName, String eDescription, MailAttachmentDisposition aDisposition) throws Exception {
        int ret = ACBrMailLib.INSTANCE.MAIL_AddAttachment(getHandle(), toUTF8(eFileName), 
                toUTF8(eDescription), aDisposition.asInt());
        checkResult(ret);
    }
    
    public void addBody(String eBody) throws Exception {
        int ret = ACBrMailLib.INSTANCE.MAIL_AddBody(getHandle(), toUTF8(eBody));
        checkResult(ret);
    }
    
    public void addAltBody(String eAltBody) throws Exception {
        int ret = ACBrMailLib.INSTANCE.MAIL_AddAltBody(getHandle(), toUTF8(eAltBody));
        checkResult(ret);
    }
    
    public void saveToFile(String eFileName) throws Exception {
        int ret = ACBrMailLib.INSTANCE.MAIL_SaveToFile(getHandle(), toUTF8(eFileName));
        checkResult(ret);
    }
    
    public void clear() throws Exception {
        int ret = ACBrMailLib.INSTANCE.MAIL_Clear(getHandle());
        checkResult(ret);
    }
    
    public void send() throws Exception {
        int ret = ACBrMailLib.INSTANCE.MAIL_Send(getHandle());
        checkResult(ret);
    }
    
    public void ConfigImportar(String eArqConfig) throws Exception {
        
        int ret = ACBrMailLib.INSTANCE.MAIL_ConfigImportar(getHandle(), eArqConfig);
        checkResult(ret);
        
    }
    
    public String ConfigExportar() throws Exception {
		
        ByteBuffer buffer = ByteBuffer.allocate(STR_BUFFER_LEN);
        IntByReference bufferLen = new IntByReference(STR_BUFFER_LEN);

        int ret = ACBrMailLib.INSTANCE.MAIL_ConfigExportar(getHandle(), buffer, bufferLen);
        checkResult(ret);

        return fromUTF8(buffer, bufferLen.getValue());
		
    }
    
    @Override
    protected void UltimoRetorno(ByteBuffer buffer, IntByReference bufferLen) {
        ACBrMailLib.INSTANCE.MAIL_UltimoRetorno(getHandle(), buffer, bufferLen);
    }
}
